Modules/_ctypes/_ctypes.c

```diff
@@ -1606,48 +1525,37 @@ PyCArrayType_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
     return NULL;
 }
 
-PyTypeObject PyCArrayType_Type = {
-    PyVarObject_HEAD_INIT(NULL, 0)
-    "_ctypes.PyCArrayType",                     /* tp_name */
-    0,                                          /* tp_basicsize */
-    0,                                          /* tp_itemsize */
-    0,                                          /* tp_dealloc */
-    0,                                          /* tp_vectorcall_offset */
-    0,                                          /* tp_getattr */
-    0,                                          /* tp_setattr */
-    0,                                          /* tp_as_async */
-    0,                                          /* tp_repr */
-    0,                                          /* tp_as_number */
-    &CDataType_as_sequence,                     /* tp_as_sequence */
-    0,                                          /* tp_as_mapping */
-    0,                                          /* tp_hash */
-    0,                                          /* tp_call */
-    0,                                          /* tp_str */
-    0,                                          /* tp_getattro */
-    0,                                          /* tp_setattro */
-    0,                                          /* tp_as_buffer */
-    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE, /* tp_flags */
-    PyDoc_STR("metatype for the Array Objects"), /* tp_doc */
-    0,                                          /* tp_traverse */
-    0,                                          /* tp_clear */
-    0,                                          /* tp_richcompare */
-    0,                                          /* tp_weaklistoffset */
-    0,                                          /* tp_iter */
-    0,                                          /* tp_iternext */
-    CDataType_methods,                          /* tp_methods */
-    0,                                          /* tp_members */
-    0,                                          /* tp_getset */
-    0,                                          /* tp_base */
-    0,                                          /* tp_dict */
-    0,                                          /* tp_descr_get */
-    0,                                          /* tp_descr_set */
-    0,                                          /* tp_dictoffset */
-    0,                                          /* tp_init */
-    0,                                          /* tp_alloc */
-    PyCArrayType_new,                                   /* tp_new */
-    0,                                          /* tp_free */
+static int
+PyCArrayType_traverse(PyTypeObject *self, visitproc visit, void *arg)
+{
+    Py_VISIT(Py_TYPE(self));
+    return PyType_Type.tp_traverse((PyObject *)self, visit, arg);
```

Comment:
`PyCArrayType_Type` and `PyCSimpleType_Type` can use `CDataType_clear/traverse`, as they now have the GC protocol. If we care about the overhead of unnecessary visits to `StgDict->proto` (non-container), a comment would be nice?

Reply:
Your comments in those functions will be helpful when converting the `PyCStgDict` type.

Please check me: 
```
1) PyCStructType_Type
2) UnionType_Type
3) PyCPointerType_Type
4) PyCArrayType_Type
5) PyCSimpleType_Type
```
They are not `StgDictObject`, but their dict objects in the `tp_dict` slot is replaced with `StgDictObject` by using `Py_SETREF()`.  Previously, (4)(5) did not have `Py_TPFLAGS_HAVE_GC`, since the `stgdict->proto` member is not a container type (e.g. `str`).  And we cannot visit/clear `StgDictObject` itself in the `CDataType_traverse/clear()` as the dict is cached (reused).

Do you think (1)(2)(3) are `StgDictObject`?  They use `CDataType`'s traverse and clear functions.

Reply:
Correction: `StgDictObject` itself (`self->tp_dict`) is handled by `PyType_Type.tp_*(self)`. However, current `PyCStgDict_Type` does not have `Py_TPFLAGS_HAVE_GC`, so at least the `proto` needs to be visited here.

Reply:
Ah, sorry, I made a mistake in my first comment. I've understood your idea, and the codes have been updated. Thank you for pointing it out!